prophet-ressa/src/gen_ressa.rs

```diff
@@ -0,0 +1,96 @@
+use prophet_ressa_minify::try_minify_ressa;
+use source_code_parser::ressa::NodePattern;
+use source_code_parser::{ressa::Indexable, Language, ModuleComponent};
+use std::collections::HashSet;
+use std::fs::DirEntry;
+
+use crate::Error;
+
+/// Use the langages in the provided LAAST to determine and load the needed ReSSAs
+pub fn extract_ressas(
+    ast: &Vec<ModuleComponent>,
+    ressa_dir: &str,
+) -> Result<Vec<NodePattern>, Error> {
+    // Get subdirectories
+    let dirs = get_subdirs(ressa_dir)?;
+
+    // Find languages
+    let langs = ast
+        .iter()
+        .flat_map(|module| find_languages(module as &dyn Indexable))
+        .collect::<HashSet<Language>>();
+
+    // Create ReSSAs for the languages in the project
+    let mut ressas = vec![];
+    let minified_ressas = dirs
+        .into_iter()
+        .flat_map(|entry| {
+            let name = entry.file_name();
+            let name = name.to_str()?;
+            let lang = to_lang(name)?;
+            Some((name.to_string(), entry.path().is_dir(), lang))
+        })
+        .flat_map(|(name, is_dir, lang)| {
+            if langs.contains(&lang) {
+                if is_dir {
+                    return Some(try_minify_ressa(name));
+                } else {
+                    tracing::warn!("{:?} not a directory, cannot add {:?} ReSSA", name, lang);
+                }
+            }
+            None
+        });
+
+    // Verify no errors, abort if error
+    for entry in minified_ressas {
+        match entry {
+            Ok(ressa) => ressas.push(ressa),
+            Err(err) => return Err(Error::MinifyError(err)),
+        }
+    }
+
+    // Flatten into one vector and return
+    Ok(ressas.into_iter().flatten().collect())
+}
+
+/// Retrieve the subdirectories of the directory named by the provided string
+fn get_subdirs(ressa_dir: &str) -> Result<Vec<DirEntry>, Error> {
+    // Validate can check provided directory
+    let read_dir = match std::fs::read_dir(&ressa_dir) {
```

Comment:
Honestly I don't think this first version here is necessarily any better since we're trying to log an error here
```rust
    let read_dir = std::fs::read_dir(ressa_dir).map_or_else(
        |err| {
            tracing::warn!("Could not read directory: {:?}", err);
            Err(Error::IOError(err))
        },
        Ok,
    )?;
```
Though I think it might make more sense to log the error where get_subdirs is called since there's other errors we're not logging here, so if we weren't logging in this function it'd be as simple as this
```rust
    let read_dir = std::fs::read_dir(ressa_dir).map_err(Error::IOError)?;
```

If Result had something akin to .inspect() like on iterators then this would be much nicer but I can't seem to find anything like that